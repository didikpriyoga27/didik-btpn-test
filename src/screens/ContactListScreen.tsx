import {NavigationProp, useNavigation} from '@react-navigation/native';
import {useQuery} from '@tanstack/react-query';
import axios from 'axios';
import React, {useCallback, useMemo} from 'react';
import {
  ActivityIndicator,
  FlatList,
  RefreshControl,
  TouchableOpacity,
} from 'react-native';
import ContactItem from '../components/ContactItem';
import {StackParamList} from '../navigation/StackParamList';
import PlusIcon from '../shared/assets/svg/PlusIcon';
import BaseLayout from '../shared/components/BaseLayout';
import Text from '../shared/components/Text';
import View from '../shared/components/View';
import {Contact, baseUrl} from '../slices/contactSlice';
import Header from '../shared/components/Header';

type Response = {
  data: {
    data: Contact[];
  };
};

const ContactListScreen = () => {
  const {navigate} = useNavigation<NavigationProp<StackParamList>>();

  const fetchContacts = (): Promise<Response> => axios.get(baseUrl);

  const {data, isLoading, refetch, isRefetching} = useQuery({
    queryKey: ['contacts'],
    queryFn: fetchContacts,
  });

  const contacts: Contact[] = useMemo(
    () => data?.data?.data || [],
    [data?.data?.data],
  );

  const sortedData = useMemo(() => {
    if (contacts?.length) {
      return contacts?.sort((a, b) => {
        if (a.firstName < b.firstName) {
          return -1;
        }
        if (a.firstName > b.firstName) {
          return 1;
        }
        return 0;
      });
    }
    return [];
  }, [contacts]);

  const renderItem = useCallback((props: {item: Contact}) => {
    return <ContactItem {...props} />;
  }, []);

  if (isLoading) {
    return (
      <View className="flex-1 items-center justify-center">
        <ActivityIndicator size={'large'} />
      </View>
    );
  }

  return (
    <BaseLayout>
      <Header title={'Contact List'} />
      <FlatList
        data={sortedData}
        className="flex-1 bg-white dark:bg-slate-800 p-4"
        refreshControl={
          <RefreshControl refreshing={isRefetching} onRefresh={refetch} />
        }
        renderItem={renderItem}
      />
      <TouchableOpacity
        testID={'ButtonAddNewContact'}
        onPress={() => navigate('ContactDetailScreen', {contactId: ''})}
        className={'absolute bottom-8 right-4'}>
        <View className="bg-teal-700 p-4 rounded-full flex-row items-center">
          <PlusIcon />
          <Text className="text-white font-semibold">Add New Contact</Text>
        </View>
      </TouchableOpacity>
    </BaseLayout>
  );
};

export default ContactListScreen;
