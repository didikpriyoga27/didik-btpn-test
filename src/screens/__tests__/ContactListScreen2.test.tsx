import {useQuery} from '@tanstack/react-query';
import {render} from '@testing-library/react-native';
import React from 'react';

import ContactListScreen from '../ContactListScreen';

jest.useRealTimers();

describe('@screens/ContactListScreen isEmpty', () => {
  // @ts-ignore
  useQuery.mockReturnValue({
    data: {
      data: {
        data: [],
      },
    },
  });

  const component = () => render(<ContactListScreen />);

  test('snapshot: component', async () => {
    const {toJSON} = component();
    expect(toJSON()).toMatchSnapshot();
  });
});
