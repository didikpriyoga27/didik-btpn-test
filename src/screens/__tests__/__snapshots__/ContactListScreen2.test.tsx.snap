// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@screens/ContactListScreen isEmpty snapshot: component 1`] = `
<View
  style={
    [
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "zIndex": 10,
          },
          {
            "width": "100%",
          },
          {
            "justifyContent": "flex-end",
          },
          {
            "elevation": 4,
            "height": 48,
            "marginTop": -0,
          },
        ],
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          [
            {
              "flexDirection": "row",
            },
            {
              "alignItems": "center",
            },
            {
              "paddingLeft": 24,
              "paddingRight": 24,
            },
            {
              "height": 48,
            },
          ],
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000",
            },
            [
              {
                "position": "absolute",
              },
              {
                "left": 0,
              },
              {
                "right": 0,
              },
              {
                "flexBasis": "0%",
                "flexGrow": 1,
                "flexShrink": 1,
              },
              {
                "textAlign": "center",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
            ],
          ]
        }
      >
        Contact List
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={<RefreshControlMock />}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "backgroundColor": "#fff",
        },
        {
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
      ]
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "bottom": 32,
        "opacity": 1,
        "position": "absolute",
        "right": 16,
      }
    }
    testID="ButtonAddNewContact"
  >
    <View
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          [
            {
              "backgroundColor": "#0f766e",
            },
            {
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
            },
            {
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
            },
            {
              "flexDirection": "row",
            },
            {
              "alignItems": "center",
            },
          ],
        ]
      }
    >
      <RNSVGSvgView
        bbHeight={25}
        bbWidth={24}
        fill="none"
        focusable={false}
        height={25}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "flex": 0,
              "height": 25,
              "width": 24,
            },
          ]
        }
        width={24}
      >
        <RNSVGGroup
          fill={null}
          propList={
            [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M11 11.5v-6h2v6h6v2h-6v6h-2v-6H5v-2h6Z"
            fill={
              {
                "payload": 4294967295,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
      <Text
        style={
          [
            {
              "color": "#000",
            },
            [
              {
                "color": "#fff",
              },
              {
                "fontWeight": "600",
              },
            ],
          ]
        }
      >
        Add New Contact
      </Text>
    </View>
  </View>
</View>
`;
